### ISSUE TITLE
v2 beta: Rejecting all orders still results in a partial capture

### ISSUE THREAD

**Issue by christophersjchow** [stripe/stripe-mirakl-connector#52]
We're not sure why this is happening but when we reject all orders we end up with a partial capture of funds in Stripe. I'm not able to determine why its exactly $12.09.

<img width="647" alt="Screen Shot 2021-01-22 at 2 34 30 pm" src="https://user-images.githubusercontent.com/47544496/105443092-0fe35e00-5cbf-11eb-87e2-6074de7f6d63.png">

In this case 2 products were purchased. Here is the OR11 response:

```json
{
  "orders" : [ {
    "acceptance_decision_date" : "2021-01-22T03:06:21Z",
    "can_cancel" : false,
    "can_evaluate" : false,
    "can_shop_ship" : false,
    "channel" : null,
    "commercial_id" : "6208346295",
    "created_date" : "2021-01-22T03:05:19Z",
    "currency_iso_code" : "AUD",
    "customer" : {
      "billing_address" : {
        "city" : "Barangaroo",
        "company" : null,
        "country" : "Australia",
        "country_iso_code" : "AUS",
        "firstname" : "Justine",
        "lastname" : "Timbersaw",
        "state" : "New South Wales",
        "street_1" : "1221 Sesame Street",
        "street_2" : "",
        "zip_code" : "2000"
      },
      "civility" : null,
      "customer_id" : "jtimbs@gmail.com",
      "email" : "jtimbs@gmail.com",
      "firstname" : "Justine",
      "lastname" : "Timbersaw",
      "locale" : null,
      "shipping_address" : {
        "additional_info" : null,
        "city" : "Barangaroo",
        "company" : null,
        "country" : "Australia",
        "country_iso_code" : "AUS",
        "firstname" : "Justine",
        "internal_additional_info" : null,
        "lastname" : "Timbersaw",
        "state" : "New South Wales",
        "street_1" : "1221 Sesame Street",
        "street_2" : "",
        "zip_code" : "2000"
      }
    },
    "customer_debited_date" : null,
    "delivery_date" : {
      "earliest" : "2021-01-30T03:05:19.307Z",
      "latest" : "2021-02-03T03:05:19.307Z"
    },
    "fulfillment" : {
      "center" : {
        "code" : "DEFAULT"
      }
    },
    "has_customer_message" : false,
    "has_incident" : false,
    "has_invoice" : false,
    "last_updated_date" : "2021-01-22T03:06:21Z",
    "leadtime_to_ship" : 5,
    "order_additional_fields" : [ ],
    "order_id" : "6208346295-A",
    "order_lines" : [ {
      "can_open_incident" : false,
      "can_refund" : false,
      "cancelations" : [ ],
      "category_code" : "beauty-fragrances-body-sprays",
      "category_label" : "Body Sprays",
      "commission_fee" : 9.60,
      "commission_rate_vat" : 10.0000,
      "commission_taxes" : [ {
        "amount" : 0.96,
        "code" : "TAXDEFAULT",
        "rate" : 10.0000
      } ],
      "commission_vat" : 0.96,
      "created_date" : "2021-01-22T03:05:19Z",
      "debited_date" : null,
      "description" : null,
      "last_updated_date" : "2021-01-22T03:06:21Z",
      "offer_id" : 3504,
      "offer_sku" : "S2222",
      "offer_state_code" : "11",
      "order_line_additional_fields" : [ {
        "code" : "destination",
        "type" : "LIST",
        "value" : "Parramatta"
      } ],
      "order_line_id" : "6208346295-A-1",
      "order_line_index" : 1,
      "order_line_state" : "REFUSED",
      "order_line_state_reason_code" : "REFUSED",
      "order_line_state_reason_label" : "Rejected",
      "price" : 109.09,
      "price_additional_info" : null,
      "price_unit" : 36.36,
      "product_medias" : [ {
        "media_url" : "/media/product/image/8a949d10-53d5-43af-ac6d-1d9bf8925986",
        "mime_type" : "JPG",
        "type" : "LARGE"
      }, {
        "media_url" : "/media/product/image/aa3b2447-6608-4fa0-9dd0-06316e15c01e",
        "mime_type" : "JPG",
        "type" : "SMALL"
      }, {
        "media_url" : "/media/product/image/d785e7a5-6b20-432c-aced-af9b9aaf8c34",
        "mime_type" : "JPG",
        "type" : "MEDIUM"
      } ],
      "product_sku" : "79748f7a-49f3-4b93-960d-762cc0e47885",
      "product_title" : "Dunhill Desire Blue Cologne",
      "promotions" : [ ],
      "quantity" : 3,
      "received_date" : null,
      "refunds" : [ ],
      "shipped_date" : null,
      "shipping_price" : 0.00,
      "shipping_price_additional_unit" : null,
      "shipping_price_unit" : null,
      "shipping_taxes" : [ ],
      "taxes" : [ {
        "amount" : 10.91,
        "code" : "AU_GST",
        "tax_calculation_rule" : "PROPORTIONAL_TO_AMOUNT"
      } ],
      "total_commission" : 10.56,
      "total_price" : 109.09
    } ],
    "order_state" : "REFUSED",
    "order_state_reason_code" : null,
    "order_state_reason_label" : null,
    "order_tax_mode" : "TAX_EXCLUDED",
    "paymentType" : null,
    "payment_type" : null,
    "payment_workflow" : "PAY_ON_ACCEPTANCE",
    "price" : 109.09,
    "promotions" : {
      "applied_promotions" : [ ],
      "total_deduced_amount" : 0
    },
    "quote_id" : null,
    "shipping_carrier_code" : null,
    "shipping_company" : null,
    "shipping_deadline" : "2021-01-27T03:05:19.307Z",
    "shipping_price" : 0.00,
    "shipping_pudo_id" : null,
    "shipping_tracking" : null,
    "shipping_tracking_url" : null,
    "shipping_type_code" : "SD0001",
    "shipping_type_label" : "Standard Delivery by Westfield",
    "shipping_zone_code" : "SHIPPING_ZONE_AU",
    "shipping_zone_label" : "Australia",
    "shop_id" : 2040,
    "shop_name" : "Try1 Store",
    "total_commission" : 10.56,
    "total_price" : 109.09
  }, {
    "acceptance_decision_date" : "2021-01-22T03:06:21Z",
    "can_cancel" : false,
    "can_evaluate" : false,
    "can_shop_ship" : false,
    "channel" : null,
    "commercial_id" : "6208346295",
    "created_date" : "2021-01-22T03:05:19Z",
    "currency_iso_code" : "AUD",
    "customer" : {
      "billing_address" : {
        "city" : "Barangaroo",
        "company" : null,
        "country" : "Australia",
        "country_iso_code" : "AUS",
        "firstname" : "Justine",
        "lastname" : "Timbersaw",
        "state" : "New South Wales",
        "street_1" : "1221 Sesame Street",
        "street_2" : "",
        "zip_code" : "2000"
      },
      "civility" : null,
      "customer_id" : "jtimbs@gmail.com",
      "email" : "jtimbs@gmail.com",
      "firstname" : "Justine",
      "lastname" : "Timbersaw",
      "locale" : null,
      "shipping_address" : {
        "additional_info" : null,
        "city" : "Barangaroo",
        "company" : null,
        "country" : "Australia",
        "country_iso_code" : "AUS",
        "firstname" : "Justine",
        "internal_additional_info" : null,
        "lastname" : "Timbersaw",
        "state" : "New South Wales",
        "street_1" : "1221 Sesame Street",
        "street_2" : "",
        "zip_code" : "2000"
      }
    },
    "customer_debited_date" : null,
    "delivery_date" : {
      "earliest" : "2021-01-30T03:05:19.307Z",
      "latest" : "2021-02-03T03:05:19.307Z"
    },
    "fulfillment" : {
      "center" : {
        "code" : "DEFAULT"
      }
    },
    "has_customer_message" : false,
    "has_incident" : false,
    "has_invoice" : false,
    "last_updated_date" : "2021-01-22T03:06:21Z",
    "leadtime_to_ship" : 5,
    "order_additional_fields" : [ ],
    "order_id" : "6208346295-B",
    "order_lines" : [ {
      "can_open_incident" : false,
      "can_refund" : false,
      "cancelations" : [ ],
      "category_code" : "westfield-direct",
      "category_label" : "Westfield Direct",
      "commission_fee" : 0.00,
      "commission_rate_vat" : 0.0000,
      "commission_taxes" : [ {
        "amount" : 0.00,
        "code" : "TAXZERO",
        "rate" : 0.0000
      } ],
      "commission_vat" : 0.00,
      "created_date" : "2021-01-22T03:05:19Z",
      "debited_date" : null,
      "description" : null,
      "last_updated_date" : "2021-01-22T03:06:21Z",
      "offer_id" : 3474,
      "offer_sku" : "S2191",
      "offer_state_code" : "11",
      "order_line_additional_fields" : [ {
        "code" : "destination",
        "type" : "LIST",
        "value" : "Parramatta"
      } ],
      "order_line_id" : "6208346295-B-1",
      "order_line_index" : 1,
      "order_line_state" : "REFUSED",
      "order_line_state_reason_code" : "REFUSED",
      "order_line_state_reason_label" : "Rejected",
      "price" : 11.77,
      "price_additional_info" : null,
      "price_unit" : 11.77,
      "product_medias" : [ ],
      "product_sku" : "c790f418-1a72-46cf-94fa-06913569ef29",
      "product_title" : "Westfield Direct Parcel Shipping",
      "promotions" : [ ],
      "quantity" : 1,
      "received_date" : null,
      "refunds" : [ ],
      "shipped_date" : null,
      "shipping_price" : 0.00,
      "shipping_price_additional_unit" : null,
      "shipping_price_unit" : null,
      "shipping_taxes" : [ ],
      "taxes" : [ {
        "amount" : 1.18,
        "code" : "AU_GST",
        "tax_calculation_rule" : "PROPORTIONAL_TO_AMOUNT"
      } ],
      "total_commission" : 0.00,
      "total_price" : 11.77
    } ],
    "order_state" : "REFUSED",
    "order_state_reason_code" : null,
    "order_state_reason_label" : null,
    "order_tax_mode" : "TAX_EXCLUDED",
    "paymentType" : null,
    "payment_type" : null,
    "payment_workflow" : "PAY_ON_ACCEPTANCE",
    "price" : 11.77,
    "promotions" : {
      "applied_promotions" : [ ],
      "total_deduced_amount" : 0
    },
    "quote_id" : null,
    "shipping_carrier_code" : null,
    "shipping_company" : null,
    "shipping_deadline" : "2021-01-27T03:05:19.307Z",
    "shipping_price" : 0.00,
    "shipping_pudo_id" : null,
    "shipping_tracking" : null,
    "shipping_tracking_url" : null,
    "shipping_type_code" : "SD0001",
    "shipping_type_label" : "Standard Delivery by Westfield",
    "shipping_zone_code" : "SHIPPING_ZONE_AU",
    "shipping_zone_label" : "Australia",
    "shop_id" : 2059,
    "shop_name" : "Westfield Direct",
    "total_commission" : 0.00,
    "total_price" : 11.77
  } ],
  "total_count" : 2
}
```

**Comment by christophersjchow** [comment]
Ok the cause of this is because the connector references `total_price` but does not take into account the `taxes` field. So we are left with an amount after the deduction logic occurs and we have the same issue in #51 where `$finalAmount > 0` so the remaining amount is captured.

Ideally this logic could be short circuited by simply iterating over the `order_state` for each logistic order and if all of them `REFUSED` or `CANCELLED` the `payment_mapping` should be cancelled.

**Comment by christophersjchow** [comment]
Wether to look at `taxes` field can be determined by looking at the `order_tax_mode` field. If this field is `TAX_EXCLUDED` add the values in `taxes` otherwise if it's `TAX_INCLUDED` ignore the `taxes` attribute.

**Comment by christophersjchow** [comment]
Last note, the happy path works, where all orders are accepted as the connector then completely avoids performing the adding/subtracting of order prices. This issue only occurs for partially accepted/refused orders.

**Comment by youssef-stripe** [comment]
Thanks for flagging @christophersjchow. We do need to add shipping_taxes and taxes to the amount deducted if the order is refused / cancelled. We will push a fix as soon as possible.

**Comment by youssef-stripe** [comment]
This should be fixed by 7a1f99f. Let me know if you still find discrepancies in the capture flow.
